File: flutter/packages/measure_flutter/lib/src/http/http_data.dart
import 'package:json_annotation/json_annotation.dart';

import '../serialization/json_serializable.dart';

part 'http_data.g.dart';

@JsonSerializable()
class HttpData implements JsonSerialized {
  /// The complete URL of the request.
  final String url;

  /// HTTP method, like get, post, put, etc. In lowercase.
  final String method;

  /// HTTP response code. Example: 200, 401, 500, etc.
  @JsonKey(name: "status_code")
  final int? statusCode;

  /// The uptime at which the http call started, in milliseconds.
  @JsonKey(name: "start_time")
  final int? startTime;

  /// The uptime at which the http call ended, in milliseconds.
  @JsonKey(name: "end_time")
  final int? endTime;

  /// The reason for the failure. Typically the IOException class name.
  @JsonKey(name: "failure_reason")
  final String? failureReason;

  /// The description of the failure. Typically the IOException message.
  @JsonKey(name: "failure_description")
  final String? failureDescription;

  /// The request headers.
  @JsonKey(name: "request_headers")
  final Map<String, String>? requestHeaders;

  /// The response headers.
  @JsonKey(name: "response_headers")
  final Map<String, String>? responseHeaders;

  /// The request body.
  @JsonKey(name: "request_body")
  final String? requestBody;

  /// The response body.
  @JsonKey(name: "response_body")
  final String? responseBody;

  /// The name of the client that sent the request.
  final String client;

  /// The number of bytes sent in the request body.
  @JsonKey(name: "bytes_sent")
  final int? bytesSent;

  /// The number of bytes received in the response body.
  @JsonKey(name: "bytes_received")
  final int? bytesReceived;

  /// The time taken for DNS resolution, in milliseconds.
  @JsonKey(name: "dns_duration")
  final int? dnsDuration;

  /// The time taken for TLS handshake, in milliseconds.
  @JsonKey(name: "tls_duration")
  final int? tlsDuration;

  /// The time taken to send the request, in milliseconds.
  @JsonKey(name: "request_send_duration")
  final int? requestSendDuration;

  /// The time taken to read the response, in milliseconds.
  @JsonKey(name: "response_read_duration")
  final int? responseReadDuration;

  /// Whether the request failed before receiving a response from the server.
  @JsonKey(name: "is_client_error")
  final bool? isClientError;

  /// Whether the request failed due to a timeout.
  @JsonKey(name: "is_timeout")
  final bool? isTimeout;

  HttpData({
    required this.url,
    required this.method,
    required this.client,
    required this.startTime,
    this.statusCode,
    this.endTime,
    this.failureReason,
    this.failureDescription,
    this.requestHeaders,
    this.responseHeaders,
    this.requestBody,
    this.responseBody,
    this.bytesSent,
    this.bytesReceived,
    this.dnsDuration,
    this.tlsDuration,
    this.requestSendDuration,
    this.responseReadDuration,
    this.isClientError,
    this.isTimeout,
  });

  @override
  Map<String, dynamic> toJson() => _$HttpDataToJson(this);

  factory HttpData.fromJson(Map<String, dynamic> json) =>
      _$HttpDataFromJson(json);
}

File: ios/Sources/MeasureSDK/Swift/Http/HttpData.swift
//
//  HttpData.swift
//  MeasureSDK
//
//  Created by Adwin Ross on 28/11/24.
//

import Foundation

struct HttpData: Codable {
    /// The complete URL of the request.
    let url: String

    /// HTTP method, like get, post, put, etc. In lowercase.
    let method: String

    /// HTTP response code. Example: 200, 401, 500, etc.
    let statusCode: Int?

    /// The uptime at which the HTTP call started, in milliseconds.
    let startTime: UnsignedNumber?

    /// The uptime at which the HTTP call ended, in milliseconds.
    let endTime: UnsignedNumber?

    /// The reason for the failure. Typically the error class name.
    let failureReason: String?

    /// The description of the failure. Typically the error message.
    let failureDescription: String?

    /// The request headers.
    var requestHeaders: [String: String]?

    /// The response headers.
    var responseHeaders: [String: String]?

    /// The request body.
    var requestBody: String?

    /// The response body.
    var responseBody: String?

    /// The name of the client that sent the request.
    let client: String

    /// The number of bytes sent in the request body.
    let bytesSent: Int64?

    /// The number of bytes received in the response body.
    let bytesReceived: Int64?

    /// The time taken for DNS resolution, in milliseconds.
    let dnsDuration: Int64?

    /// The time taken for TLS handshake, in milliseconds.
    let tlsDuration: Int64?

    /// The time taken to send the request, in milliseconds.
    let requestSendDuration: Int64?

    /// The time taken to read the response, in milliseconds.
    let responseReadDuration: Int64?

    /// Whether the request failed before receiving a response from the server.
    let isClientError: Bool?

    /// Whether the request failed due to a timeout.
    let isTimeout: Bool?

    enum CodingKeys: String, CodingKey {
        case url
        case method
        case statusCode = "status_code"
        case startTime = "start_time"
        case endTime = "end_time"
        case failureReason = "failure_reason"
        case failureDescription = "failure_description"
        case requestHeaders = "request_headers"
        case responseHeaders = "response_headers"
        case requestBody = "request_body"
        case responseBody = "response_body"
        case client
        case bytesSent = "bytes_sent"
        case bytesReceived = "bytes_received"
        case dnsDuration = "dns_duration"
        case tlsDuration = "tls_duration"
        case requestSendDuration = "request_send_duration"
        case responseReadDuration = "response_read_duration"
        case isClientError = "is_client_error"
        case isTimeout = "is_timeout"
    }
}
